### PR TITLE
Plfm 7546.1

### PR DESCRIPTION
**Pull request by lukemoenning**

### CODE REVIEW

File: lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/stats/StatGeneratorIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,200 @@
+package org.sagebionetworks.table.cluster.stats;
+
+import static org.junit.Assert.assertEquals;
```
Comment (john-hill) [inline lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/stats/StatGeneratorIntegrationTest.java]: need a jupiter

File: lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/stats/StatGeneratorIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+	private Long stringListColMaxSize = 97L;
+	private Long doubleColMaxSize = 131L;
+	private Long integerColMaxSize = 157L;
+	
+	
+	@BeforeEach
+	public void before() throws ParseException {
+		setupTableAndColumnMapper();
+	}
+
```
Comment (john-hill) [inline lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/stats/StatGeneratorIntegrationTest.java:57]: missing tests around null inputs

File: lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/stats/ColumnReferenceGenerator.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+import org.sagebionetworks.table.cluster.TableAndColumnMapper;
+import org.sagebionetworks.table.cluster.columntranslation.ColumnTranslationReference;
+import org.sagebionetworks.table.query.model.ColumnReference;
+
+public class ColumnReferenceGenerator implements StatGeneratorInteface<ColumnReference> {
+
+	@Override
+	public Optional<ElementStats> generate(ColumnReference element, TableAndColumnMapper tableAndColumnMapper) {
+		Optional<ColumnTranslationReference> ctrOptional = tableAndColumnMapper.lookupColumnReference(element);
```
Comment (john-hill) [inline lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/stats/ColumnReferenceGenerator.java:13]: if(option.isEmpty(){
return Optional.empty();
}
ElementStats stats =  optional.get();
rest of the code use stats

File: lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/stats/ElementStats.java [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+		if(two == null) {
+			return one;
+		}
+		return Math.max(one, two);
+	}
+	
+	public static <T> T lastNonNull(T one, T two) {
+		return two != null ? two : one;
+	}
+
```
Comment (john-hill) [inline lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/stats/ElementStats.java:56]: can add a new function:
Builder createClonedBuilder(){
pre-populate with all of this values.
}

File: lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/stats/StatGeneratorIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+		
+		assertEquals(expected, statGenerator.generate(
+				new TableQueryParser(valueExpression).valueExpression(), mockTableAndColumnMapper));
+	}
+	
+	@Test
+	public void testGenerateWithConcatWithCount() throws ParseException {
+		String valueExpression = "CONCAT(foo, '1', COUNT(foo))";
+		Optional<ElementStats> expected = Optional.of(ElementStats.builder()
+				.setMaximumSize(fooMaxSize + 1L + ColumnConstants.MAX_INTEGER_BYTES_AS_STRING).build());
```
Comment (john-hill) [inline lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/stats/StatGeneratorIntegrationTest.java]: Will need a constant for:
	public static final int MAX_INTEGER_CHARS_AS_STRING = Long.toString(-Long.MAX_VALUE).length();

File: lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/stats/StatGeneratorIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+		
+		assertEquals(expected, statGenerator.generate(
+				new TableQueryParser(valueExpression).valueExpression(), mockTableAndColumnMapper));
+	}
+	
+	@Test
+	public void testGenerateWithAverage() throws ParseException {
+		String valueExpression = "AVG(foo)";
+		Optional<ElementStats> expected = Optional.of(ElementStats.builder()
+				.setMaximumSize(Long.valueOf(ColumnConstants.MAX_DOUBLE_BYTES_AS_STRING)).build());
```
Comment (john-hill) [inline lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/stats/StatGeneratorIntegrationTest.java]: we want number of chars not bytes

File: lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/stats/MySqlFunctionGenerator.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+		switch (element.getFunctionName()) {
+			case CONCAT:
+				for (ValueExpression valueExpression : element.getParameterValues()) {
+					Optional<ElementStats> childStats = new StatGenerator().generate(valueExpression, tableAndColumnMapper);
+					
+					if (childStats.isEmpty()) {
+						return Optional.empty();
+					}
+					
+					elementStats = ElementStats.generateSumStats(elementStats, childStats.get());
```
Comment (john-hill) [inline lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/stats/MySqlFunctionGenerator.java]: clone and only change the maxSize.

File: lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/stats/NumericValueExpressionGenerator.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+		if (termStats.isEmpty()) {
+			return Optional.empty();
+		}
+		
+		ElementStats termPrimeStats = ElementStats.builder().build();
+		
+		for (TermPrime termPrime : numericValueExpression.getPrimeList()) {
+			Optional<ElementStats> child = generator.generate(termPrime, tableAndColumnMapper);
+			if (!child.isEmpty()) {
+				termPrimeStats = ElementStats.generateSumStats(termPrimeStats, child.get());
```
Comment (john-hill) [inline lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/stats/NumericValueExpressionGenerator.java]: I should always be safe to return the size of a double

Add test for:
concat(foo, aNum1 + aNum2) where 1&2 are numeric columns.

File: lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/stats/SetFunctionSpecificationGenerator.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+import org.sagebionetworks.table.query.model.SetFunctionSpecification;
+import org.sagebionetworks.table.query.model.ValueExpression;
+
+public class SetFunctionSpecificationGenerator implements StatGeneratorInteface<SetFunctionSpecification> {
+
+	@Override
+	public Optional<ElementStats> generate(SetFunctionSpecification element, TableAndColumnMapper tableAndColumnMapper) {
+		switch (element.getSetFunctionType().getFunctionReturnType()) {
+			case LONG:
+				return buildStats(ColumnConstants.MAX_INTEGER_BYTES_AS_STRING);
```
Comment (john-hill) [inline lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/stats/SetFunctionSpecificationGenerator.java]: use the number of chars not bytes

File: lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/stats/StatGeneratorInteface.java [hunk 10]
```diff
@@ -0,0 +1,11 @@
+package org.sagebionetworks.table.cluster.stats;
+
+import java.util.Optional;
+
+import org.sagebionetworks.table.cluster.TableAndColumnMapper;
+import org.sagebionetworks.table.query.model.Element;
+
+@FunctionalInterface
+public interface StatGeneratorInteface<T extends Element> {
+	Optional<ElementStats> generate(T element, TableAndColumnMapper tableAndColumnMapper);
```
Comment (john-hill) [inline lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/stats/StatGeneratorInteface.java:26]: comments what is expected of an implementer.

File: lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/stats/TermPrimeGenerator.java [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+
+import java.util.Optional;
+
+import org.sagebionetworks.table.cluster.TableAndColumnMapper;
+import org.sagebionetworks.table.query.model.TermPrime;
+
+public class TermPrimeGenerator implements StatGeneratorInteface<TermPrime> {
+
+	@Override
+	public Optional<ElementStats> generate(TermPrime element, TableAndColumnMapper tableAndColumnMapper) {
```
Comment (john-hill) [inline lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/stats/TermPrimeGenerator.java]: remove if not used after other changes.

File: lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/SQLTranslatorUtilsTest.java [hunk 12]
```diff
@@ -3976,28 +3969,91 @@ public void testGetSchemaOfDerivedColumnWithDerivedWithAverage() throws ParseExc
...
+				.setColumnType(ColumnType.STRING)
+				.setMaximumSize(50L)
+				.setId(null);
+		
+		assertEquals(exp, SQLTranslatorUtils.getSchemaOfDerivedColumn(dc, mapper));
+	}
+	
 	@Test
 	public void testGetSchemaOfDerivedColumnWithDerivedWithConcat() throws ParseException {
-		QueryExpression rootModel = new TableQueryParser("select concat(foo,\"-\", bar) AS \"concat\" from syn123").queryExpression();
```
Comment (john-hill) [inline lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/SQLTranslatorUtilsTest.java:3987]: concat(foo, '-', bar)

File: lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/stats/SetFunctionSpecificationGeneratorUnitTest.java [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+	@Mock
+	private TableAndColumnMapper mockTableAndColumnMapper;
+	
+	@InjectMocks
+	private SetFunctionSpecificationGenerator mockSetFunctionSpecificationGenerator;
+	
+	
+	@Test
+	public void testGenerateWithLong() throws ParseException {
+		SetFunctionSpecification element = new SetFunctionSpecification(true);
```
Comment (john-hill) [inline lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/stats/SetFunctionSpecificationGeneratorUnitTest.java]: use the parser to create your elements

File: lib/lib-table-query/src/main/java/org/sagebionetworks/table/query/model/TermPrime.java [hunk 14]
```diff
@@ -24,6 +24,14 @@ public TermPrime(ArithmeticOperator operator, Term term) {
 		this.term = term;
 	}
 
+	public ArithmeticOperator getOperator() {
```
Comment (john-hill) [inline lib/lib-table-query/src/main/java/org/sagebionetworks/table/query/model/TermPrime.java]: remove if unused

File: lib/lib-table-query/src/main/java/org/sagebionetworks/table/query/model/ValueExpressionList.java [hunk 15]
```diff
@@ -15,6 +15,10 @@ public ValueExpressionList() {
 		this.list = new ArrayList<>();
 	}
 	
+	public List<ValueExpression> getList() {
```
Comment (john-hill) [inline lib/lib-table-query/src/main/java/org/sagebionetworks/table/query/model/ValueExpressionList.java]: remove unused.

File: lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/stats/ElementStats.java [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+public class ElementStats {
+	private final Long maximumSize;
+	private final Long maxListLength;
+	private final String defaultValue;
+	private final FacetType facetType;
+	private final List<String> enumValues;
+	private final List<JsonSubColumnModel> jsonSubColumns;
+	
+	
+	private ElementStats(Long maximumSize, Long maxListLength, String defaultValue, FacetType facetType,
```
Comment (john-hill) [inline lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/stats/ElementStats.java]: need tests for constructor (covers the builder) and cloneBuilder()

File: lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/stats/NumericValueExpressionGenerator.java [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+
+public class NumericValueExpressionGenerator implements StatGeneratorInteface<NumericValueExpression> {
+
+	@Override
+	public Optional<ElementStats> generate(NumericValueExpression element, TableAndColumnMapper tableAndColumnMapper) {
+		if (element.getPrimeList().size() == 0) {
+			return new StatGenerator().generate(element.getTerm(), tableAndColumnMapper);
+		}
+		
+		return Optional.of(ElementStats.builder()
```
Comment (john-hill) [inline lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/stats/NumericValueExpressionGenerator.java:17]: this one is not obvious but is correct, so comments can help

File: lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/stats/StatGenerator.java [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+		
+		if (element instanceof Term) {
+			return generate(((Term) element).getFactor(), tableAndColumnMapper);
+		}
+		
+		if (element instanceof Factor) {
+			return generate(((Factor) element).getNumericPrimary(), tableAndColumnMapper);
+		}
+		
+		if (element instanceof SimpleBranch) {
```
Comment (john-hill) [inline lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/stats/StatGenerator.java:61]: this is the one case that is special so a comment about what we are doing here can co a long way.

File: lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/SQLTranslatorUtils.java [hunk 19]
```diff
@@ -1367,68 +1372,40 @@ public static List<ColumnModel> getSchemaOfSelect(SelectList selectList,
...
-		result.setFacetType(facetType);
-		result.setDefaultValue(defaultValue);
-		result.setEnumValues(enumValues);
-		result.setJsonSubColumns(jsonSubColumns);
-		result.setId(null);
+		
+		ColumnModel result = new ColumnModel()
+				.setColumnType(columnType)
+				.setName(selectColumn.getName())
+				.setMaximumSize(elementStats.getMaximumSize())
```
Comment (john-hill) [inline lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/SQLTranslatorUtils.java]: we should only set the max size for strings and string lists.

File: lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/SQLTranslatorUtils.java [hunk 20]
```diff
@@ -1367,68 +1372,40 @@ public static List<ColumnModel> getSchemaOfSelect(SelectList selectList,
...
+				.setMaximumListLength(elementStats.getMaxListLength())
```
Comment (john-hill) [inline lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/SQLTranslatorUtils.java]: we should only show the maxListLength for lists.

**Review by john-hill — CHANGES_REQUESTED** [review]

File: lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/SQLTranslatorUtilsTest.java [hunk 21]
```diff
@@ -3956,7 +3949,7 @@ public void testGetSchemaOfDerivedColumnWithDerivedWithCount() throws ParseExcep
 		ColumnModel expected = new ColumnModel();
 		expected.setName("count");
 		expected.setColumnType(ColumnType.INTEGER);
-		expected.setMaximumSize(null);
+		expected.setMaximumSize(Long.valueOf(ColumnConstants.MAX_INTEGER_BYTES_AS_STRING));
```
Comment (john-hill) [inline lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/SQLTranslatorUtilsTest.java]: should still be null

File: lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/SQLTranslatorUtilsTest.java [hunk 22]
```diff
@@ -3976,28 +3969,91 @@ public void testGetSchemaOfDerivedColumnWithDerivedWithAverage() throws ParseExc
 		ColumnModel expected = new ColumnModel();
 		expected.setName("AVG(id)");
 		expected.setColumnType(ColumnType.DOUBLE);
-		expected.setMaximumSize(null);
+		expected.setMaximumSize(Long.valueOf(ColumnConstants.MAX_DOUBLE_CHARACTERS_AS_STRING));
 		expected.setId(null);
```
Comment (john-hill) [inline lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/SQLTranslatorUtilsTest.java]: still be null

File: lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/SQLTranslatorUtils.java [hunk 23]
```diff
@@ -1367,68 +1372,40 @@ public static List<ColumnModel> getSchemaOfSelect(SelectList selectList,
 	 */
 	public static ColumnModel getSchemaOfDerivedColumn(DerivedColumn derivedColumn,
 			TableAndColumnMapper tableAndColumnMapper) {
-		// the SelectColumn provides a starting name and type.
+		// The SelectColumn provides a starting name and type.
 		SelectColumn selectColumn = getSelectColumns(derivedColumn, tableAndColumnMapper);
-		Long maximumSize = null;
-		Long maxListLength = null;
-		List<String> enumValues = null;
```
Comment (john-hill) [inline lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/SQLTranslatorUtils.java:1374]: we only need the stats to include maxSize and maxListSize.

**Review by john-hill — CHANGES_REQUESTED** [review]

File: lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/SQLTranslatorUtils.java [hunk 24]
```diff
@@ -1367,67 +1368,32 @@ public static List<ColumnModel> getSchemaOfSelect(SelectList selectList,
...
-	 * @param newValue
-	 * @return
-	 */
-	public static Long addLongsWithNull(Long currentValue, Long newValue) {
-		if(currentValue == null) {
-			return newValue;
-		}
-		if(newValue == null) {
-			return currentValue;
+			result.setDefaultValue(cm.getDefaultValue())
```
Comment (john-hill) [inline lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/SQLTranslatorUtils.java]: Foo clone = EntityFactory.createEntityFromJSON(EntityFactory.createJSONObjectForEntity(original), Foo.class));

File: lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/stats/StatGenerator.java [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+		if (element instanceof Term) {
+			return generate(((Term) element).getFactor(), tableAndColumnMapper);
+		}
+		
+		if (element instanceof Factor) {
+			return generate(((Factor) element).getNumericPrimary(), tableAndColumnMapper);
+		}
+		
+		// Catch all for SimpleBranches
+		// Should be the last if statement as some of the implemented cases above may also be of SimpleBranch
```
Comment (john-hill) [inline lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/stats/StatGenerator.java]: use multi-line comment

File: lib/models/src/main/java/org/sagebionetworks/repo/model/table/ColumnConstants.java [hunk 26]
```diff
@@ -61,6 +61,18 @@ public class ColumnConstants {
 	 */
 	public static final int MAX_USER_ID_BYTES_AS_STRING = MAX_INTEGER_BYTES_AS_STRING;
 	
+	/**
+	 * The maximum number of characters of a integer (a long in java terms) when represented as a string.
+	 */
+	public static final int MAX_INTEGER_CHARACTERS_AS_STRING = Long.toString(-Long.MIN_VALUE).length();
```
Comment (john-hill) [inline lib/models/src/main/java/org/sagebionetworks/repo/model/table/ColumnConstants.java]: remove '-'

**Review by john-hill — CHANGES_REQUESTED** [review]

**Review by john-hill — APPROVED** [review]
